dataactvalidator/scripts/load_program_activity.py

```diff
@@ -20,54 +29,179 @@ def get_program_activity_file(base_path):
             base_path: directory of domain config files
 
         Returns:
-            Returns the file path for the pa file either on S3 or locally
+            the file path for the pa file either on S3 or locally
     """
-
     if CONFIG_BROKER["use_aws"]:
-        s3connection = boto.s3.connect_to_region(CONFIG_BROKER['aws_region'])
-        s3bucket = s3connection.lookup(CONFIG_BROKER['sf_133_bucket'])
-        pa_file = s3bucket.get_key("program_activity.csv").generate_url(expires_in=600)
+        s3_object = get_s3_object()
+        response = s3_object.get(PA_SUB_KEY+PA_FILE_NAME)
+        pa_file = io.BytesIO(response['Body'].read())
     else:
-        pa_file = os.path.join(base_path, "program_activity.csv")
+        pa_file = os.path.join(base_path, PA_FILE_NAME)
 
     return pa_file
 
 
+def get_s3_object():
+    """ Gets the s3 connection
+
+        Returns:
+            s3bucket connection object
+    """
+    if CONFIG_BROKER["use_aws"]:
+        s3 = boto3.resource('s3')
+        s3_object = s3.Object(PA_BUCKET, PA_SUB_KEY + PA_FILE_NAME)
+        return s3_object
+    raise Exception("Config specifies not to use AWS, cannot call S3 bucket.")
+
+
+def get_date_of_current_pa_upload(base_path):
+    """ Gets the last time the file was uploaded to S3, or alternatively the last time the local
+        file was modified.
+
+        Args:
+            base_path: directory of domain config files
+
+        Returns:
+            DateTime object
+    """
+    if CONFIG_BROKER["use_aws"]:
+        last_uploaded = boto3.client('s3').head_object(Bucket=PA_BUCKET, Key=PA_SUB_KEY+PA_FILE_NAME)['LastModified']
+    else:
+        pa_file = get_program_activity_file(base_path)
+        last_uploaded = datetime.datetime.utcfromtimestamp(os.path.getmtime(pa_file))
+    return last_uploaded
+
+
+def get_stored_pa_last_upload():
+    """ Gets last recorded timestamp from last time file was processed.
+
+        Returns:
+            Upload date of most recent file we have recorded (Datetime object)
+    """
+    sess = GlobalDB.db().session
+    last_stored_obj = sess.query(ExternalDataLoadDate).join(ExternalDataType).filter(
```

Comment:
Do we need to do this join? We have a constant in `lookups.py` called `EXTERNAL_DATA_TYPE` that we can use to get the ID, which is what's stored in `ExternalDataLoadDate` so there's really no need to make a join.